spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/RewriteAppendArcticTable.scala

```diff
@@ -41,42 +43,50 @@ case class RewriteAppendArcticTable(spark: SparkSession) extends Rule[LogicalPla
     case a @ AppendData(r: DataSourceV2Relation, query, writeOptions, _) if isArcticRelation(r) =>
       val arcticRelation = asTableRelation(r)
       val upsertWrite = arcticRelation.table.asUpsertWrite
-      val (newQuery, options) = if (upsertWrite.appendAsUpsert()) {
+      val (newQuery, options, projections) = if (upsertWrite.appendAsUpsert()) {
         val upsertQuery = rewriteAppendAsUpsertQuery(r, query)
+        val insertQuery = Project(Seq(Alias(Literal(UPDATE_OPERATION), OPERATION_COLUMN)()) ++ upsertQuery.output, upsertQuery)
+        val projections = buildInsertProjections(insertQuery, r.output, isUpsert = true)
         val upsertOptions = writeOptions + (WriteMode.WRITE_MODE_KEY -> WriteMode.UPSERT.mode)
-        (upsertQuery, upsertOptions)
+        (insertQuery, upsertOptions, projections)
       } else {
-        (query, writeOptions)
+        val insertQuery = Project(Seq(Alias(Literal(INSERT_OPERATION), OPERATION_COLUMN)()) ++ query.output, query)
+        val projections = buildInsertProjections(insertQuery, r.output, isUpsert = false)
+        (insertQuery, writeOptions, projections)
       }
+
       arcticRelation.table match {
         case tbl: ArcticSparkTable =>
           if (tbl.table().isKeyedTable) {
-            val validateQuery = buildValidatePrimaryKeyDuplication(r, query)
             if (checkDuplicatesEnabled()) {
-              AppendArcticData(arcticRelation, newQuery, validateQuery, options)
+              val writeOption = options + ("optimize.enabled" -> "true")
+              val validateQuery = buildValidatePrimaryKeyDuplication(r, query)
+              val checkDataQuery = QueryWithConstraintCheck(newQuery, validateQuery)
+              ArcticRowLevelWrite(arcticRelation, checkDataQuery, writeOption, projections)
             } else {
-              ReplaceArcticData(arcticRelation, newQuery, options)
+              ArcticRowLevelWrite(arcticRelation, newQuery, options, projections)
             }
           } else {
             a
           }
       }
```

Comment:
I think those codes are too complex.  maybe we could split into different case ?

case AppendData if needUpsert() =>

Reply:
I'll take the advice，I will optimize this code